### PR TITLE
Ability to add single switch flow into diverse group Closes #2072

### LINKED ISSUE #2072: Ability to add single switch flow into diverse group
Now it is forbidden for technical reasons
Related #1231

### PR DESCRIPTION
**Pull request by Fedichk**
Closes #2072

### CODE REVIEW

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidator.java [hunk 1]
```diff
@@ -399,7 +399,7 @@ private void checkOneSwitchFlowConflict(FlowEndpoint source, FlowEndpoint destin
 
     @VisibleForTesting
     void checkDiverseFlow(RequestedFlow targetFlow) throws InvalidFlowException {
-        if (targetFlow.getSrcSwitch().equals(targetFlow.getDestSwitch())) {
+        if (targetFlow.isOneSwitchFlow()) {
             throw new InvalidFlowException("Couldn't add one-switch flow into diverse group",
```
Comment (niksv) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidator.java:403]: Actually you can add one switch flow into a diverse group. Yes, long flow wouldn't be rerouted immediately but, it long flow will be rerouted later - it will try to stay away from one switch flow. So I think this `if` can be removed

File: src-java/flowhs-topology/flowhs-storm-topology/src/test/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidatorTest.java [hunk 2]
```diff
@@ -231,46 +242,79 @@ public void checkForEncapsulationTypeRequirementCorrectTypeTest() throws Invalid
...
+                .flowId("oneSwitchFlow")
+                .srcSwitch(singleSwitch)
+                .srcPort(PORT_1)
+                .destSwitch(singleSwitch)
+                .destPort(11)
+                .diverseGroupId(DIVERSE_FLOW_ID)
+                .build();
+
+        when(flowRepository.findById(DIVERSE_FLOW_ID)).thenReturn(Optional.of(oneSwitchFlow));
+        flowValidator.checkDiverseFlow(flow);
```
Comment (dmitrii-beliakov) [inline src-java/flowhs-topology/flowhs-storm-topology/src/test/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidatorTest.java:299]: It looks strange that there is no a single assertion. You can at least make sure that the flows exist. Alternatively, you can use `assertThatNoException()` idiom from assetj or you can add  `@Test(expected = None.class)` to make it explicit that no exception is expected.

File: src-java/kilda-pce/src/test/java/org/openkilda/pce/finder/DiversityPathFinderTest.java [hunk 3]
```diff
@@ -393,6 +393,91 @@ public void shouldChooseLongerPath() throws RecoverableException, UnroutableFlow
...
+        isls.addAll(getBidirectionalIsls(switchD, 7, switchE, 8));
+        isls.addAll(getBidirectionalIsls(switchE, 9, switchA, 10));
+
+        FlowPath forwardPath = FlowPath.builder()
+                .srcSwitch(switchE)
+                .destSwitch(switchE)
+                .pathId(FORWARD_PATH_ID)
+                .segments(Collections.emptyList())
+                .build();
+        when(flowPathRepository.findById(FORWARD_PATH_ID)).thenReturn(java.util.Optional.of(forwardPath));
```
Comment (dmitrii-beliakov) [inline src-java/kilda-pce/src/test/java/org/openkilda/pce/finder/DiversityPathFinderTest.java:417]: just curious why the fully qualified name is used in this test? Any conflict with some other Optional?

File: src-java/kilda-pce/src/test/java/org/openkilda/pce/impl/AvailableNetworkTest.java [hunk 4]
```diff
@@ -35,11 +37,14 @@
 import org.hamcrest.Matchers;
 import org.junit.Test;
 
+import java.util.Collections;
 import java.util.Set;
 import java.util.UUID;
 import java.util.stream.IntStream;
 
 public class AvailableNetworkTest {
+    private static final int cost = 700;
```
Comment (dmitrii-beliakov) [inline src-java/kilda-pce/src/test/java/org/openkilda/pce/impl/AvailableNetworkTest.java:46]: Constants are usually in upper case, like the weight function below

File: src-java/kilda-pce/src/test/java/org/openkilda/pce/impl/AvailableNetworkTest.java [hunk 5]
```diff
@@ -222,7 +211,7 @@ public void shouldSetEqualCostForPairedLinks() {
     }
 
     @Test
-    public void shouldCreateSymmetricOutgoingAndIncomming() {
+    public void createSymmetricOutgoingAndIncomingTest() {
```
Comment (dmitrii-beliakov) [inline src-java/kilda-pce/src/test/java/org/openkilda/pce/impl/AvailableNetworkTest.java:214]: Could you please add a word link to the method's name? It's outgoing and incoming links

**Review by dmitrii-beliakov — APPROVED** [review]

File: src-java/kilda-pce/src/test/java/org/openkilda/pce/finder/DiversityPathFinderTest.java (hunk 3 above)
Comment (Fedichk) [reply src-java/kilda-pce/src/test/java/org/openkilda/pce/finder/DiversityPathFinderTest.java:417]: Good catch, thanks! Will fix it.

File: src-java/kilda-pce/src/test/java/org/openkilda/pce/impl/AvailableNetworkTest.java (hunk 4 above)
Comment (Fedichk) [reply src-java/kilda-pce/src/test/java/org/openkilda/pce/impl/AvailableNetworkTest.java:46]: You are right, I will fix it. Thanks!

File: src-java/kilda-pce/src/test/java/org/openkilda/pce/impl/AvailableNetworkTest.java (hunk 5 above)
Comment (Fedichk) [reply src-java/kilda-pce/src/test/java/org/openkilda/pce/impl/AvailableNetworkTest.java:214]: Sure, will do. Thanks!

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidator.java (hunk 1 above)
Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidator.java:403]: I was thinking, that such behavior would confuse a user, but delayed rerouting works fine, so, fixed. Thank you!

**Review by niksv — APPROVED** [review]

**Comment by pkazlenka** [comment]
@Fedichk , the feature works itself, but the overlapping stats, obtained from `/api/v1/{flow-id}/path` returns incorrect overlapping stats in `diverseGroup` and `otherFlows` objects for one-switch flows. Please, fix it.
Also, it would be great if you move test `two one-switch flows in the same diversity group` to unit level. This is a quite exotic setup (AFAIK) and we shouldn't waste minutes to check the one on functional test level.
TIA.

**Comment by Fedichk** [comment]
> @Fedichk , the feature works itself, but the overlapping stats, obtained from `/api/v1/{flow-id}/path` returns incorrect overlapping stats in `diverseGroup` and `otherFlows` objects for one-switch flows. Please, fix it. Also, it would be great if you move test `two one-switch flows in the same diversity group` to unit level. This is a quite exotic setup (AFAIK) and we shouldn't waste minutes to check the one on functional test level. TIA.

Added logic and tests

File: src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java [hunk 6]
```diff
@@ -56,25 +57,46 @@ public class IntersectionComputer {
...
+    }
 
-        otherEdges = paths.stream()
-                .filter(e -> !e.getFlow().getFlowId().equals(targetFlowId))
-                .flatMap(path -> path.getSegments().stream())
-                .collect(Collectors.groupingBy(
-                        PathSegment::getPathId,
-                        Collectors.mapping(Edge::fromPathSegment, Collectors.toSet())
-                ));
+    private void handleAnotherPath(FlowPath path) {
```
Comment (dmitrii-beliakov) [inline src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java]: Could you explain what are the target and the other switches? I don't get the purpose of this class.
I think the naming could be improved here. `Another` means one more, as we can say in the bar `another beer!` meaning we ask +1 beer. `Other` means different, we can say in the bar `other beer!` meaning we don't like this one, we want some different beer instead of this one.

File: src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java [hunk 7]
```diff
@@ -201,6 +206,19 @@ public static List<PathSegment> calculatePathIntersectionFromDest(Collection<Flo
...
+    private OverlappingSegmentsStats computeIntersectionCounters(Set<Edge> edges, Set<SwitchId> switches) {
+        int edgesOverlap = Sets.intersection(edges, targetPathEdges).size();
+        int switchesOverlap = Sets.intersection(switches, targetPathSwitches).size();
+        return new OverlappingSegmentsStats(edgesOverlap,
+                switchesOverlap,
+                percent(edgesOverlap, targetPathEdges.size()),
+                percent(switchesOverlap, targetPathSwitches.size()));
+    }
+
+    private int percent(int n, int from) {
```
Comment (dmitrii-beliakov) [inline src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java:211]: `n` and `from` are no so informative. Why from? meaning total?

File: src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java [hunk 8]
```diff
@@ -201,6 +206,19 @@ public static List<PathSegment> calculatePathIntersectionFromDest(Collection<Flo
...
+        return (int) ((n * 100.0f) / from);
```
Comment (dmitrii-beliakov) [inline src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java:212]: n = 1, from = 1000, gives 0%. This 0% is really different from 0% when n = 0. 
Also, division by 0 is not handled here.

File: src-java/flowhs-topology/flowhs-storm-topology/src/test/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidatorTest.java [hunk 9]
```diff
@@ -258,46 +269,98 @@ public void checkForEncapsulationTypeRequirementCorrectTypeTest() throws Invalid
         flowValidator.checkForEncapsulationTypeRequirement(SRC_ENDPOINT, properties, TRANSIT_VLAN);
     }
 
-    private RequestedFlow getTestRequestWithMaxLatencyAndMaxLatencyTier2(Long maxLatency, Long maxLatencyTier2) {
-        return RequestedFlow.builder()
+    @Test(expected = InvalidFlowException.class)
+    public void failIfMaxLatencyTier2HigherThanMaxLatencyTest() throws InvalidFlowException {
+        RequestedFlow flow = getTestRequestWithMaxLatencyAndMaxLatencyTier2((long) 1000, (long) 500);
```
Comment (dmitrii-beliakov) [inline src-java/flowhs-topology/flowhs-storm-topology/src/test/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidatorTest.java:274]: Some people consider casting as an anti-pattern. For literals you can use L: `1000L`, for other cases there is parsing methods or valueOf in the wrapper classes

File: src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java (hunk 6 above)
Comment (Fedichk) [reply src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java]: Regarding naming 'Another' means 'an' + 'other' and applies only to one object. For many objects, it will be 'other'.
Regarding purpose - this class calculates overlapping between flows. Target flow is like the "main flow" to which we are comparing other flows.

File: src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java (hunk 7 above)
Comment (Fedichk) [reply src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java:211]: It was there, I didn't rename it, but yeah you are right, I will rename it.

File: src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java (hunk 8 above)
Comment (Fedichk) [reply src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java:212]: It is the same situation, it is the old code, I will fix it. Thanks!

File: src-java/flowhs-topology/flowhs-storm-topology/src/test/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidatorTest.java (hunk 9 above)
Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/test/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidatorTest.java:274]: The old code will update. Thanks!

**Review by niksv — CHANGES_REQUESTED** [review]
Code looks good, but I've found failed test case:
1. Create flow1 from switches A to B (without diverse)
2. Create flow2 from switches A to C (without diverse)
3. Update flow2: src switch A, dst switch A, "diverse_flow_id": "flow1"

Expected:
updated flow

Actual:
 "Couldn't add one-switch flow into diverse group"
 
 @pkazlenka Please add corresponding func tests for update/patch

File: src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java (hunk 8 above)
Comment (Fedichk) [reply src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java:212]: Regarding our configuration max allowed depth is 35, so the situation when you 1/1000 is unreal in our current environment.

File: src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java (hunk 7 above)
Comment (Fedichk) [reply src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java]: Fixed.

File: src-java/flowhs-topology/flowhs-storm-topology/src/test/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidatorTest.java (hunk 9 above)
Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/test/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidatorTest.java]: Fixed.

**Comment by Fedichk** [comment]
> Code looks good, but I've found failed test case:
> 
> 1. Create flow1 from switch A to B (without diverse)
> 2. Create flow2 from switch A to C (without diverse)
> 3. Update flow2: src switch A, dst switch A, "diverse_flow_id": "flow1"
> 
> Expected: updated flow
> 
> Actual: "Couldn't add one-switch flow into diverse group"
> 
> @pkazlenka Please add corresponding func tests for update/patch

The logic is updated. I think it should work now.

**Review by niksv — APPROVED** [review]
